torch/_functorch/_activation_checkpointing/rematerialize_nodes_with_ac_annotations_pass.py

```diff
@@ -0,0 +1,140 @@
+    # This is what eliminate_dead_code() calls when is_impure_node=None
+    impure_random = True
+    if torch._guards.TracingContext.try_get():
+        impure_random = torch._inductor.config.fallback_random
+    return node.is_impure(impure_random)
+
+
+def _is_backward_node(node: fx.Node) -> bool:
+    """Check if node is in backward region via annotation"""
+    return node.meta.get("custom", {}).get("backward") is not None
+
+
+def rematerialize_nodes_with_ac_annotations(gm: fx.GraphModule) -> fx.GraphModule:
+    """
+    Duplicate checkpointed nodes for backward use. DCE removes unused forward versions.
+    """
+    if not has_recomputable_ops(gm):
+        return gm
+
+    if has_recomputable_rng_ops(gm):
+        raise RuntimeError(
+            "Activation checkpoint rematerializing in `forward-loss-backward` graph does not support RNG ops "
+            "in checkpointed regions. Please move RNG operations outside "
+            "of checkpoint regions, or use joint graph mode (where partitioner handles RNG)."
+        )
+
+    # Use partitioner pass to normalize AC node tags.
+    gm = cleanup_recompute_tags(gm, is_default_partition=False)
+
+    if not config.unsafe_allow_optimization_of_collectives:
+        force_save_collectives(gm)
+
+    force_save_bw_mutation_src(gm)
+
+    # Find backward boundary and build ordering
+    bwd_start: int | None = None
+    order = {}
+    for idx, node in enumerate(gm.graph.nodes):
+        order[node] = idx
+        if _is_backward_node(node) and bwd_start is None:
+            bwd_start = idx
+
+    if bwd_start is None:
+        raise RuntimeError(
+            "We are trying to rematerialize AC nodes in the backward region, but we could not find any backward nodes. "
+            'This is likely because you forgot to annotate your backward region with fx.traceback.annotate({"backward": 0}) '
+        )
+
+    new_graph = fx.Graph()
+    env: dict[fx.Node, fx.Node] = {}
+    recomputed_nodes: dict[fx.Node, fx.Node] = {}
+
+    def remat_input(x):
+        # fx.Node can be int or float
+        if not isinstance(x, fx.Node):
+            return x
+        return recomputed_nodes.get(x, env[x])
+
+    for node in gm.graph.find_nodes(op="placeholder"):
+        env[node] = new_graph.node_copy(node, lambda x: env[x])
+
+    def gather_checkpointed_deps(node: fx.Node, visited: set) -> None:
+        if node in visited or node in recomputed_nodes:
+            return
+        visited.add(node)
+        for inp in node.all_input_nodes:
+            if must_recompute(inp):
+                gather_checkpointed_deps(inp, visited)
+
+    # Insert forward nodes
+    for node in list(gm.graph.nodes)[:bwd_start]:
+        if node.op == "placeholder":
+            continue
+        env[node] = new_graph.node_copy(node, lambda x: env[x])
+
+    # Insert backward nodes, assumption here is that
+    # you already annotated your backward region with fx.traceback.annotate({"backward": 0})
```

Comment:
Maybe move the note about annotation assumption to the above where we computed bwd_start